wagtail/contrib/modeladmin/tests/test_simple_modeladmin.py

```diff
@@ -854,6 +856,18 @@ def test_delete_post_permitted(self):
         response = self.client.post("/admin/modeladmintest/book/delete/2/")
         self.assertRedirects(response, "/admin/")
 
+    def test_permission_helper(self):
+        permission_helper = PermissionHelper(Admin)
+
+        # Populate user permissions cache
+        with self.assertNumQueries(2):
+            self.assertTrue(self.user.has_perm("wagtailadmin.access_admin"))
+
+        with self.assertNumQueries(1):
```

Comment:
Issue: I would have expected to see a case in here where 0 queries were performed. If the codenames were already cached, as were the user permissions, then checking this a second time should result in 0 queries. That'd be an ideal test of this caching mechanism.

Reply:
This makes sense in that this PR only caches getting the permission codenames, `user_has_any_permissions` still ends up calling `user.has_perm()`, so that is the one query we see here.

lru_cache for `user_has_any_permissions()`?

Reply:
This PR caches the model's permissions codenames. That's what we want to test.

The lines above have nothing to do with the permission helper, it's Django fetching all the user's permissions.
The first time we interact with the `pemission_helper` is when we do: `permission_helper.user_has_any_permissions(self.user)`.
This will fetch codenames and populate the cache.
The following line, we do: `permission_helper.user_can_list(self.user)` which now uses the cache.

In the commits history, you can see that I've added a test first and it was failing. Adding the cache fixes it.
Does that make sense?

Reply:
this was nitpicking territory :)
happy with it as it is.

note to self: `permission_helper.user_can_list()` is plain `permission_helper.user_has_any_permission()` under the hood
https://github.com/wagtail/wagtail/blob/d4f1ec47a96d993fbb297fab279f4987ecfeb15a/wagtail/contrib/modeladmin/helpers/permission.py#L60-L65